include/odp/api/spec/packet.h

```diff
@@ -71,6 +71,51 @@ extern "C" {
   * Packet is red
   */
 
+/**
+ * Protocol
+ */
+typedef enum odp_proto_t {
+	/** No protocol defined */
+	ODP_PROTO_NONE = 0,
+
+	/** Ethernet (including VLAN) */
+	ODP_PROTO_ETH,
+
+	/** IP (including IPv4 and IPv6) */
```

Comment:
Maybe it would be better to split this into separate IPv4 and IPv6 packets. It would be an error to pass IPv6 packet with ethtype (or IP tunnel type) being set to IPv4.

Reply:
That might be overly complicated since until a decrypted tunnel mode IPsec packet is parsed you don't know whether it's IPv4 or IPv6. It's parsing that makes that determination.

Reply:
@Bill-Fischofer-Linaro In IPsec case Next Header field will contain 4 for IPv4 and 41 for IPv6.

Reply:
First bits of an IP header marks the version. So, it would be trivial for both app and implementation to read the version from the data. Common IP define is easier for application when a burst of packets may contain both v4 and v6 mixed. Application does not need to sort  packets into two arrays (one for v4 and other for v6) but just pass the entire array for parsing.

There are three ways to define the enumeration: IP, IPv4+IPv6, IP+IPv4+IPv6. I'm OK with any of those. IPv4+IPv6 would be a bit more robust since version information comes from two sources.

Reply:
@psavol yes. I just wanted to focus on cases, when passing packet with wrong protocol is an error. E.g. IPv6 packet inside IPsec packet with NH = 4. So it is not a question of selecting proper L3 parser, but rather a question of nailing down error/malicious packets.

Reply:
@psavol Also for multi-packet parsing, we can change `proto` to be an array, allowing applications to easily intermix IPv4 and IPv6 packets.

Reply:
@Bill-Fischofer-Linaro Because it is a parsing error, if lower layer provides us IP version that does not correspond to the in-packet version. We should detect that, rather than silently parsing this header.